File: source java/utilities/position/Point.java
package utilities.position;

import com.sun.glass.events.KeyEvent;
import utilities.savingsystem.Loader;
import utilities.savingsystem.Saver;

public class Point extends Position
{
    private static Point[] neighbours_ = {new Point(-1, 0), new Point(1, 0), new Point(0, 1), new Point(0, - 1)};
    
    public int x;
    public int y;
    
    public Point()
    {
        this.x = this.y = 0;
    }
    
    public Point(int x, int y)
    {
        this.x = x;
        this.y = y;
    }
    
    @Override
    public int Distance(Position other)
    {
        if(!(other instanceof Point))
            return -1;
        
        Point otherPoint = (Point)(other);
        //if(metrics == WorldMode.SQUARE)
            return (Math.abs(this.x - otherPoint.x) + Math.abs(this.y - otherPoint.y));
       // else if(metrics == WorldMode.HEXAGONS)
          //  return (Math.abs(this.x - other.x) + Math.abs(this.x + this.y - other.x - other.y) + Math.abs(this.y - other.y)) / 2;
       // else
         //   return -1;
    }
    
    @Override
    public Point Add(Position other)
    {
        Point otherPoint = (Point)(other);
        return new Point(x + otherPoint.x, y + otherPoint.y);
    }
    
    @Override
    public void Save(Saver saver)
    {
        saver.Print(Integer.toString(x));
        saver.Print(Integer.toString(y));
    }
    
    @Override
    public Point Load(Loader loader)
    {
        x = loader.NextInt();
        y = loader.NextInt();
        return this;
    }
    
    @Override
    public int hashCode()
    {
        return (x + y) * (x + y + 1) / 2 + y;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        
        final Point other = (Point) obj;
        return (other.x == this.x && other.y == this.y);
    }
    
    @Override
    public String toString()
    {
        return "Point (" + x + ", " + y + ")";
    }

    @Override
    public Position GetNeighbouringCloserPosition(Position destination) 
    {
        if(!(destination instanceof Point))
                System.out.println("Gdzieś popełniłem bląd");
        
        int distance = this.Distance(destination);
        
        for(Point neighbour : neighbours_)
        {
            int newDistance = (this.Add(neighbour)).Distance(destination);
            if(distance > newDistance)
                return this.Add(neighbour);
        }
        
        return this;
    }

    @Override
    public boolean IsInBoundaries(int width, int height) 
    {
       return x >= 0 && x < width && y >= 0 && y < height;
    }

    @Override
    public Point[] GetNeighbours() 
    {
        Point neighbours[] = new Point[neighbours_.length];
        for(int i = 0; i < neighbours_.length; i++)
            neighbours[i] = this.Add(neighbours_[i]);
        return neighbours;
    }
    
    @Override
    public Point SetRandom(int width, int height)
    {
        x = generator.nextInt(width);
        y = generator.nextInt(height);
        return this;
    }
    
    @Override
    public Point GetMovement(int keyCode)
    {
        switch(keyCode)
        {
            case KeyEvent.VK_UP:
                return this.Add(new Point(0, -1));
            case KeyEvent.VK_DOWN:
                return this.Add(new Point(0, 1));
            case KeyEvent.VK_LEFT:
                return this.Add(new Point(-1, 0));
            case KeyEvent.VK_RIGHT:
                return this.Add(new Point(1, 0));
        }
        
        return null;
    }

    @Override
    public Position GetRandomNeighbour() {
        return this.Add(neighbours_[generator.nextInt(neighbours_.length)]);
    }
}

File: source python/Utilities/Position/Point.py
import random
import pygame


class Point:

    def __init__(self, x = 0, y = 0):
        self.x = x
        self.y = y
    
    def __hash__(self):
        return int((self.x + self.y) * (self.x + self.y + 1) / 2 + self.y)

    def __str__(self):
        return "Point (" + str(self.x) + ", " + str(self.y) + ")"

    def __eq__(self, other):
        if not isinstance(other, Point):
            return False
        return (self.x, self.y) == (other.x, other.y)

    def __ne__(self, other):
        return not (self == other)
    
    def __add__(self, other):
        return Point(self.x + other.x, self.y + other.y)

    @staticmethod
    def GetRandom(width, height):
        return Point(random.randint(0, width - 1), random.randint(0, height - 1))
    
    def Distance(self, point):
        return abs(self.x - point.x) + abs(self.y - point.y)

    def IsInBoundary(self, width, height):
        return self.x >= 0 and self.x < width and self.y >= 0 and self.y < height

    def GetRandomNeighbour(self):
        return self + Point.Neighbours()[random.randint(0, len(Point.Neighbours()) - 1)]

    def GetAllNeighbours(self):
        absoluteNeighbours = Point.Neighbours()
        result = list()
        for absoluteNeighbour in absoluteNeighbours:
            result.append(self + absoluteNeighbour)
        return result

    def GetNeighbouringCloserPosition(self, position):
        distance = self.Distance(position)
        if distance == 0:
            return self

        neighbours = self.GetAllNeighbours()
        for neighbour in neighbours:
            currentDistance = neighbour.Distance(position)
            if currentDistance < distance:
                return neighbour
        return self

    def GetMovement(self, key):
        if key == pygame.K_UP:
            return self + Point.Neighbours()[2]
        elif key == pygame.K_DOWN:
            return self + Point.Neighbours()[3]
        elif key == pygame.K_LEFT:
            return self + Point.Neighbours()[0]
        elif key == pygame.K_RIGHT:
            return self + Point.Neighbours()[1]
        else:
            return None

    @staticmethod
    def Neighbours():
        if not hasattr(Point.Neighbours, "neighbours"):
            Point.Neighbours.__neighbours = [Point(-1, 0), Point(1, 0), Point(0, -1), Point(0, 1)]
        return Point.Neighbours.__neighbours

    def Save(self, saver):
        saver.Insert(self.x)
        saver.Insert(self.y)

    @staticmethod
    def Load(loader):
        return Point(int(loader.Next()), int(loader.Next()))







